Material/Material2D/DuncanSelig.cpp

```diff
@@ -0,0 +1,273 @@
+
+int DuncanSelig::update_trial_status(const vec& t_strain) {
+    incre_strain = (trial_strain = t_strain) - current_strain;
+
+    if(norm(incre_strain) <= datum::eps) return SUANPAN_SUCCESS;
+
+    trial_history = current_history;
+    auto& max_dev_stress = trial_history(0);
+
+    // assuming elastic loading/unloading
+    trial_stress = current_stress + (trial_stiffness = initial_stiffness) * incre_strain;
+
+    // if elastic response exceeds the maximum deviatoric stress, then the material is considered to be in a plastic state
+    if(dev(trial_stress) <= max_dev_stress) return SUANPAN_SUCCESS;
+
+    vec3 net_stress = current_stress, net_strain = incre_strain;
+
+    if(dev(current_stress) < max_dev_stress) {
+        // project onto the yield surface first
+        auto elastic_portion = .5;
+        if(SUANPAN_SUCCESS != project_to_surface(elastic_portion)) return SUANPAN_FAIL;
+
+        net_stress = current_stress + initial_stiffness * incre_strain * elastic_portion;
+        net_strain = (1. - elastic_portion) * incre_strain;
+    }
+
+    auto ref_error = 0.;
+    vec3 incre;
+    mat33 jacobian;
+
+    auto counter = 0u;
+    while(true) {
+        if(max_iteration == ++counter) {
+            suanpan_error("Cannot converge within {} iterations.\n", max_iteration);
+            return SUANPAN_FAIL;
+        }
+
+        const auto [elastic, bulk, deds, dkds] = compute_moduli();
+
+        const auto factor_a = 3. * bulk * (3. * bulk + elastic);
+        const auto factor_b = 3. * bulk * (3. * bulk - elastic);
+
+        mat33 right(fill::zeros);
+        right(0, 0) = right(1, 1) = factor_a;
+        right(0, 1) = right(1, 0) = factor_b;
+        right(2, 2) = 3. * bulk * elastic;
+
+        const vec3 residual = (9. * bulk - elastic) * (trial_stress - net_stress) - right * net_strain;
+
+        const rowvec3 dfads = (18. * bulk + 3. * elastic) * dkds + 3. * bulk * deds;
+        const rowvec3 dfbds = (18. * bulk - 3. * elastic) * dkds - 3. * bulk * deds;
+
+        jacobian = (9. * bulk - elastic) * eye(3, 3) + (trial_stress - net_stress) * (9. * dkds - deds);
+        jacobian.row(0) -= net_strain(0) * dfads + net_strain(1) * dfbds;
+        jacobian.row(1) -= net_strain(0) * dfbds + net_strain(1) * dfads;
+        jacobian.row(2) -= net_strain(2) * 3. * (bulk * deds + elastic * dkds);
+
+        if(!solve(incre, jacobian, residual)) return SUANPAN_FAIL;
+
+        const auto error = inf_norm(incre);
+        if(1u == counter) ref_error = error;
+        suanpan_debug("Local iteration error: {:.5E}.\n", error / ref_error);
+        if(error < tolerance * ref_error || ((error < tolerance || inf_norm(residual) < tolerance) && counter > 5u)) {
+            if(!solve(trial_stiffness, jacobian, right)) return SUANPAN_FAIL;
+
+            max_dev_stress = dev(trial_stress);
+
+            return SUANPAN_SUCCESS;
+        }
+
+        trial_stress -= incre;
+    }
```

Comment:
This loop effectively implements the incremental constitutive relation Eq. (3a). To obtain the consistent tangent matrix, we need to take full derivative w.r.t. stress as now the elastic and bulk moduli are functions of stress. The reference implementation provided by Prof. Katona does not follow this approach, as a result, it cannot achieve the quadratic convergence rate.